hbase-shell/src/main/ruby/shell.rb

```diff
@@ -302,33 +315,13 @@ def get_workspace
       # Install all the hbase commands, constants, and instance variables @shell and @hbase. This
       # will override names that conflict with IRB methods like "help".
       export_all(hbase_receiver)
+      # make it so calling exit will hit our pass-through rather than going directly to IRB
+      hbase_receiver.send :define_singleton_method, :exit, lambda { |rc = 0|
```

Comment:
Why must this be done via runtime injection instead of defining a method on the `HBaseReceiver` class?

Reply:
Presumably because we don't have a handle to `@shell` in the receiver

Reply:
the call on line 314 to `IRB::ExtendCommandBundle` would overwrite an `exit` method defined on the `HBaseReceiver` class.

we could complicate our lives by passing a different override option to the `extend_object` call that said we didn't want to override methods already defined on the receiver, but then IRB would always print a warning that looked like this below because it also defines an exit method.

```
    print "irb: warn: can't alias #{to} from #{from}.\n"
```

Reply:
we don't have a handle to `@shell`, that's true. but if we didn't have the override issue we could e.g. keep the return code state in the receiver and pull it out when the shell needs to return it.